File: app/src/test/java/org/mifos/mobilebanking/LoanRepaymentSchedulePresenterTest.java
package org.mifos.mobilebanking;

import android.content.Context;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mifos.mobilebanking.api.DataManager;
import org.mifos.mobilebanking.models.accounts.loan.LoanWithAssociations;
import org.mifos.mobilebanking.presenters.LoanRepaymentSchedulePresenter;
import org.mifos.mobilebanking.ui.views.LoanRepaymentScheduleMvpView;
import org.mifos.mobilebanking.util.RxSchedulersOverrideRule;
import org.mifos.mobilebanking.utils.Constants;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;


import io.reactivex.Observable;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by dilpreet on 27/6/17.
 */
@RunWith(MockitoJUnitRunner.class)
public class LoanRepaymentSchedulePresenterTest {

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Mock
    Context context;

    @Mock
    DataManager dataManager;

    @Mock
    LoanRepaymentScheduleMvpView view;

    private LoanRepaymentSchedulePresenter presenter;
    private LoanWithAssociations loanWithRepaymentSchedule, loanWithEmptyRepaymentSchedule;

    @Before
    public void setUp() throws Exception {
        presenter = new LoanRepaymentSchedulePresenter(context, dataManager);
        presenter.attachView(view);

        loanWithRepaymentSchedule = FakeRemoteDataSource.getLoanAccountRepaymentSchedule();
        loanWithEmptyRepaymentSchedule = FakeRemoteDataSource.
                getLoanAccountEmptyRepaymentSchedule();

    }

    @After
    public void tearDown() throws Exception {
        presenter.detachView();
    }

    @Test
    public void testLoadLoanAccountDetails() throws Exception {
        when(dataManager.getLoanWithAssociations(Constants.REPAYMENT_SCHEDULE, 29)).
                thenReturn(Observable.just(loanWithRepaymentSchedule));

        presenter.loanLoanWithAssociations(29);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showLoanRepaymentSchedule(loanWithRepaymentSchedule);
        verify(view, never()).showEmptyRepaymentsSchedule(null);
        verify(view, never()).showError(context.
                getString(R.string.error_fetching_repayment_schedule));

    }

    @Test
    public void testLoadLoanAccountDetailsEmpty() throws Exception {
        when(dataManager.getLoanWithAssociations(Constants.REPAYMENT_SCHEDULE, 29)).
                thenReturn(Observable.just(loanWithEmptyRepaymentSchedule));

        presenter.loanLoanWithAssociations(29);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showEmptyRepaymentsSchedule(loanWithEmptyRepaymentSchedule);
        verify(view, never()).showLoanRepaymentSchedule(null);
        verify(view, never()).showError(context.
                getString(R.string.error_fetching_repayment_schedule));

    }


    @Test
    public void testLoadLoanAccountDetailsFails() throws Exception {
        when(dataManager.getLoanWithAssociations(Constants.REPAYMENT_SCHEDULE, 29)).
                thenReturn(Observable.<LoanWithAssociations>error(new RuntimeException()));

        presenter.loanLoanWithAssociations(29);

        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showError(context.getString(R.string.error_fetching_repayment_schedule));
        verify(view, never()).showEmptyRepaymentsSchedule(null);
        verify(view, never()).showLoanRepaymentSchedule(null);

    }
}

File: app/src/test/java/org/mifos/mobile/LoanRepaymentSchedulePresenterTest.kt
package org.mifos.mobile

import android.content.Context

import io.reactivex.Observable

import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import org.mifos.mobile.api.DataManager
import org.mifos.mobile.models.accounts.loan.LoanWithAssociations
import org.mifos.mobile.presenters.LoanRepaymentSchedulePresenter
import org.mifos.mobile.ui.views.LoanRepaymentScheduleMvpView
import org.mifos.mobile.util.RxSchedulersOverrideRule
import org.mifos.mobile.utils.Constants

import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

/**
 * Created by dilpreet on 27/6/17.
 */
@RunWith(MockitoJUnitRunner::class)
class LoanRepaymentSchedulePresenterTest {
    @Rule
    @JvmField
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()

    @Mock
    var context: Context? = null

    @Mock
    var dataManager: DataManager? = null

    @Mock
    var view: LoanRepaymentScheduleMvpView? = null
    private var presenter: LoanRepaymentSchedulePresenter? = null
    private var loanWithRepaymentSchedule: LoanWithAssociations? = null
    private var loanWithEmptyRepaymentSchedule: LoanWithAssociations? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        presenter = LoanRepaymentSchedulePresenter(context, dataManager!!)
        presenter?.attachView(view)
        loanWithRepaymentSchedule = FakeRemoteDataSource.loanAccountRepaymentSchedule
        loanWithEmptyRepaymentSchedule = FakeRemoteDataSource.loanAccountEmptyRepaymentSchedule
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        presenter?.detachView()
    }

    @Test
    @Throws(Exception::class)
    fun testLoadLoanAccountDetails() {
        Mockito.`when`(dataManager?.getLoanWithAssociations(Constants.REPAYMENT_SCHEDULE, 29)).thenReturn(Observable.just(loanWithRepaymentSchedule))
        presenter?.loanLoanWithAssociations(29)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showLoanRepaymentSchedule(loanWithRepaymentSchedule)
        Mockito.verify(view, Mockito.never())?.showEmptyRepaymentsSchedule(null)
        Mockito.verify(view, Mockito.never())?.showError(context?.getString(R.string.error_fetching_repayment_schedule))
    }

    @Test
    @Throws(Exception::class)
    fun testLoadLoanAccountDetailsEmpty() {
        Mockito.`when`(dataManager?.getLoanWithAssociations(Constants.REPAYMENT_SCHEDULE, 29)).thenReturn(Observable.just(loanWithEmptyRepaymentSchedule))
        presenter?.loanLoanWithAssociations(29)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showEmptyRepaymentsSchedule(loanWithEmptyRepaymentSchedule)
        Mockito.verify(view, Mockito.never())?.showLoanRepaymentSchedule(null)
        Mockito.verify(view, Mockito.never())?.showError(context?.getString(R.string.error_fetching_repayment_schedule))
    }

    @Test
    @Throws(Exception::class)
    fun testLoadLoanAccountDetailsFails() {
        Mockito.`when`(dataManager?.getLoanWithAssociations(Constants.REPAYMENT_SCHEDULE, 29)).thenReturn(Observable.error(RuntimeException()))
        presenter?.loanLoanWithAssociations(29)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showError(context?.getString(R.string.error_fetching_repayment_schedule))
        Mockito.verify(view, Mockito.never())?.showEmptyRepaymentsSchedule(null)
        Mockito.verify(view, Mockito.never())?.showLoanRepaymentSchedule(null)
    }
}